rfcs/0039-buffer-physical-layout.md

```diff
@@ -0,0 +1,522 @@
+
+  - Option 2: `BufferTransform` nodes listed in the `PrimFunc::attrs`
+    under a `"buffer_argument_transforms"` key apply to the function arguments.
+
+    Would only need a single traversal to apply.
+
+    Would require other passes to be aware of where a buffer was first
+    defined, in order to add it to the appropriate location.
+
+
+# Prior art
+[prior-art]: #prior-art
+
+- CuDNN has an [explicit enumeration of allowed input
+  formats](https://docs.nvidia.com/deeplearning/cudnn/api/index.html#cudnnTensorFormat_t),
+  which are specific to image formatting.
+
+- The reorder/split/flatten sequences is equivalent in numpy to using
+  `np.reshape` to split the logical axes, then `np.transpose` to
+  reorder them, then `np.reshape` to merge multiple axes into the N-d
+  physical axes.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+
+- What is appropriate terminology for size/shape/extent of physical
+  and logical buffers?
+
+  If Allocate/BufferStore/BufferLoad each hold a reference to the
+  buffer they act upon, then this becomes a somewhat irrelevant
+  question, as there is only one `BufferNode::shape`.
+
+  - I am partial to using "shape" both for the N-d parameters, and
+    have attempted to use it consistently through this RFC.
+  - "size" implies a 1-d buffer, which wouldn't be appropriate for
+    an N-d parameter.
+  - "extent" would be a reasonable name, but is currently used by
+    `tvm::RangeNode` to indicate a range of values that may start at
+    a non-zero value.  Since the indices for logical and physical
+    buffers both start at zero, using "extents" for the maximum
+    index would imply some offset.
+
+
+
+- How should loops over an array be handled when re-writing the shape?
+
+  To avoid memory latency issues, loops should iterate over an array
+  sequentially sequentially when possible.  Iteration that is defined
+  in terms of the logical layout may be inappropriate for the physical
+  layout.
+
+  Option 3 is preferred.
+
+  - Option 1: Do nothing, and always keep the same iteration order,
+    using the same iteration axes as defined in the compute
+    definition.
+
+    This would produce valid code, but not necessarily performant
+    code.  This can be a default behavior during development, to be
+    improved upon.
+
+  - Option 2: Automatically detect loops that are over the full extent
+    of an array in sequential order of the logical layout, and rewrite
+    to be in sequential order of the physical layout.
+
+    This would reduce the memory latency issues, but raises some
+    implementation questions.
+
+    - If a loop body references multiple tensors with different
+      physical layouts, which should define the loop iteration order?
+
+    - If a series of nested loops contains a `cache_read` or
+      `cache_write` stage, can these be recognized and reordered?
+
+  - Option 3: Expose the `reorder_split` definition to be used as part
```

Comment:
Need updates as `reorder_split` has been renamed, could you also explain how it works?

Reply:
In TE, the transformed axes would be defined in the return value from `s[A].transform_layout(...)`, so that the remainder of the TE schedule can refer to `.op.axis` of the return value.  (Updating the wording in the RFC shortly.)

In TensorIR, I *think* it would be done similarly, with `AA = T.transform(A, lambda ...)` returning the transformed buffer.  The loops in the schedule could then be written over `T.grid(*AA.shape)`.  I'm not as familiar with `tvm.script.tir`, so does that sound reasonable?

Reply:
sounds reasonable, this will allow scheduling on the transformed axes